---
title: "P/BIO381 Tutorials: Population Genomics 7: Analyzing results from Fst outlier (Bayescan) runs"
date: 'Mar 26, 2018'
output:
  prettydoc::html_pretty:
    theme: cayman
fontsize: 18pt
---
## Goals:

* **Discuss** Markov Chain Monte Carlo (mcmc) approaches to estimation of Bayesian model parameters

* **Assess** convergence of our mcmc runs in 'Bayescan' and the effects of different thinning internvals within the chains

* **Compare** the sensitivity of outlier loci from 'Bayescan' runs assuming different prior odds of selection

* **Annotate** our results to determine the functional roles of our outliers

------------------------------

### Plotting and Interpreting Results

So, you've got some output from Bayescan now....what does it all mean?

First, let's talk a little about What markov chain monte carlo (mcmc) is, and how it generated Bayesian estimates of your selection model...

The job that Bayescan and other evolutionary genetic models have is tough -- they are trying to estimate complex models often with many parameters.

Oftentimes, we don't know what the distribution should look like for these set of parameters, so we approximate the likelihood by sampling the space most consistent with the data. That's where mcmc comes in. It's a search strategy for finding regions of high likelihood, and using this to compute your posterior parameters:

![]("~/Downloads/mcmc_likelihood_fig.png")

{{ :tutorials:mcmc_likelihood_fig.png?1000 |}}

There are a couple of issues you need to be aware of when using mcmc programs. First, is that the initial start of the chain can bias your estimates if you include these values. People generally discard the first several thousand iterations as "burn-in". Second, saving iterations that are too close together in the chain generates autocorrelation in your estimates, which can also bias your parameters. The solution to this is "thinning".

Do you remember the burn-in and thinning values we used in our runs? What were they?

Most importantly -- do not blindly trust the output of ANY program. You __have__ to look at your data and results to ensure things are OK.

Now, let's get on with our Bayescan results.....

You'll notice there are a couple of different files in your output. They should looks something like this:

<cli>
[srkeller@pbio381 subset]$ ll
total 1216
-rw-r--r--. 1 srkeller users    601 Oct 27 20:59 sites_example.recode.baye_AccRte.txt
-rw-r--r--. 1 srkeller users 262301 Oct 27 21:02 sites_example.recode.baye_fst.txt
-rw-r--r--. 1 srkeller users 249247 Oct 27 15:16 sites_example.recode.bayescan
-rw-r--r--. 1 srkeller users 272641 Oct 27 21:02 sites_example.recode.baye.sel
-rw-r--r--. 1 srkeller users 445749 Oct 27 15:18 sites_example.recode.baye_Verif.txt
[srkeller@pbio381 subset]$ 
</cli>

Here's what's in these files:

  * **AccRte.txt:**    This contains information on the acceptance rate for updates during the mcmc chain.
  * **fst.txt:**       This a row for each SNP site, and contains the probability that it's under selection, the log10 Posterior Odds that the locus is selected vs. neutral, the value of alpha, the corresponding qvalue indicating the probability of a false discovery, and lastly, the Fst for that locus
  * **.sel:**           This contains the output from the mcmc chain; can be useful for plotting and ensuring your run converged (look at logL)
  * **Verif.txt:**     This is essentially a log file that contains information on your run and how the commands were interpreted by Bayescan.

The most important files are the "_fst.txt" with your selection results, and the ".sel" with the info on how the likelihood varied across your run.

We'll want to download these to our local machine for plotting in R. Here's our workflow:

  - Use 'scp' to copy the files to your local machine. You remember how, right?
  - Bring your results files into R and make plots of the change in log(Likelihood) across the run. Was the burn-in long enough? 
  - Also in R, determine if there are loci that exhibit significant evidence of selection, assuming a False Discovery Rate of 0.01
  - Make a 4 panel plot of (a) a histogram of alpha, (b) a histogram of Fst, (c) a bi-plot of alpha vs. Fst across loci, and (d) a plot of alpha vs. the -log10(q-value).
  - Create a list of candidate loci that show evidence of selection. Export from R as a tab-delimited text file for further analysis
  - Use "grep" to determine the annotation of these genes in the OTAU draft reference genome. What genes are they? Look up some of the protein family (pfam) categories they belong to.

**Here's an [[:tutorials:Detecting selection using Fst -- BAYESCAN:R script|R script]] to get you going.**
