---
title: "P/BIO381 Tutorials: Working with RNA-seq data"
date: 'January 29, 2018'
output:
  prettydoc::html_pretty:
    theme: cayman
fontsize: 18pt
---

# Learning Objectives for 1/29/18

1. To understand the Onthophagus taurus beetle biology, experimental design, and data set.
2. To understand the general work flow or "pipeline" for processing and analyzing RNAseq data.
3. To learn how to make/write a bash script and how write a script to process files in batches (*.sh files and #! should mean something to you by the end of today).
4. To visualize and interpret Illumina data quality (what is a .fastq file and what are Phred scores?).
5. To add to your growing list of bioinformatics tricks: using the `screen` command and how to test scripts with a subset of your data.



## 1. *Onthophagus taurus*

Native to the mediterranean, this dung beetle was deliberately introduced to Australia in the late 1960s with a massive, government led effort to help reduce fly pests and disease that were plaguing livestock.  In the early 1970s, the beetles were "accidentally" introduced to the eastern United States from an unknown origin.  

*The experimental design:*

- Three populations reared in a common garden for more than two generations: from the native range, Italy (IT), from Western Australia (WA), from North Carolina (NC)
- Four developmental stages: late third larval instart (L3L), pre-pupae day 1 (PP1), pupae day 1 (PD1), and adults four days after ecolsion (AD4)
- Both sexes; three individuals per sex
- 3 pops * 4 developmental stages * 2 sexes * 3 individuals = 72 samples
- Sequenced on about 7 lanes of Illumina HiSeq 2500



## 2. The "pipeline"

1. a. Visualize the quality of raw data (Program: FastQC)

   b. Clean raw data (Program: Trimmomatic)

   c. Visualize the quality of cleaned data (Program: FastQC)

2. Download reference transcriptome assembly

   *Note:* If you're working with an organism that does not have a reference assembly, this is the point when you use your cleaned reads to make a reference assembly (e.g. program: Trinity), assess it's quality (Programs: BUSCO, blastx to various databases)

3. Map (a.k.a. Align) cleaned reads from each sample to the reference assembly to generate **s**equence **a**lign**m**ent files (Program: bwa, Input: *.fastq, Output: *.sam)

4. Extract read count data from .sam files (i.e. the number of reads that map (align) to each "gene")

5. Assemble a data matrix of counts for each gene for each sample

6. Analyze count data to test for differences in gene expression



## 3. Visualize, Clean, and Visualize again

Let's begin to work with our RNAseq data!  The first step is to take a look at the quality of our data coming off the sequencer and see if we notice any problems.  

You'll each be in charge of cleaning and visualizing the left (R1) and right (R2) files from a single sample.

Your file assignment for today:

| Student | File      |
| ------- | --------- |
| Ethan   | WA_PP1_F1 |
| Jamie   | WA_PP1_F2 |
| Lucy    | WA_PP1_F3 |
| Morgan  | WA_PP1_M1 |
| Paul    | WA_PP1_M2 |
| Suraj   | WA_PP1_M3 |

You can find these in the directory `/data/project_data/beetles/rawdata/`

### Peak

Let's take a peak into a file and learn something about what a .fastq file is:

```bash
zcat YOURFILENAME | head
```

**The fast file format** has 4 lines for each read: the read identifier, the read sequence, "+", and a sequence of quality scores for each base.

[Here's a useful reference for understanding Quality (Phred) scores](http://www.drive5.com/usearch/manual/quality_score.html).  If P is the probability that a base call is an error, then:

P = 10^(–Q/10) 

Q = –10 log10(P)

***The Q score is translated to ASCII characters so that a two digit number can be represented by a single character.*** So cool! 

Typically, we accept bases with Q >= 30, which is equivelant to a 0.1% chance of error (40 is 0.01% error, 20 is 1% error, 10 is 10% error).  

*Based on the ASCII code table linked above, what kind of characters do you want to see in your quality score?* 

### Visualize

Now how can we look at the quality more systematically for all reads in the file?  We can use [the program FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) (also already installed in our `/data/popgen/` directory).

Now clean your file using the following command:

```
fastqc FILENAME.fq.gz
```

Move your .html file to your computer using the `scp`command from your machine (hint: open another terminal or putty window)

```
scp NETID@pbio381.uvm.edu:YOURDIRECTORY/YOURFILE_fastqc.html .
```

The dot means "to the present directory" or you can direct it somewhere else.

Alternatively, you can use a file tranfer program like fetch, etc.

*How does the quality look?*

### Clean

[Here's a link to the Trimmomatic program](http://www.usadellab.org/cms/index.php?page=trimmomatic) that we'll use to clean the reads for each file. The program is already installed in our `/data/popgen/` directory.

There's an example script in the `/data/scripts/` directory.  Make a directory in your directory called "scripts" and copy the bash script over, edit in vim, make it executable, and run it!

```bash
cp /data/scripts/trim_example.sh ~/scripts/ # copies the script
chmod u+x trim_example.sh                   # makes the script "executable"
./trim_example.sh  		# executes the script, or bash trim_example.sh
```

Now check the quality of one of your cleaned files using fastqc again.



